appinventor/misc/emulator-support/aiStarter.py

```diff
@@ -71,61 +95,98 @@ def ucheck():
+            f'"{PLATDIR}adb" -s {device} shell am start -a android.intent.action.VIEW -n edu.mit.appinventor.aicompanion3/.Screen1 --ez rundirect true',
+            shell=True)
+        response.headers['Access-Control-Allow-Origin'] = '*'
+        response.headers['Access-Control-Allow-Headers'] = 'origin, content-type'
+        return ''
+    except subprocess.CalledProcessError as e:
+        print(f"Problem starting companion app : status {e.returncode}\n")
+        return ''
+
 
 def checkrunning(emulator):
-    result = subprocess.check_output(PLATDIR + 'commands-for-Appinventor/adb devices', shell=True, close_fds=True)
-    lines = result.split('\n')
-    for line in lines[1:]:
-        if emulator:
-            m = re.search('^(.*emulator-[1-9]+)\t+device.*', line)
-        else:
-            if re.search('^(.*emulator-[1-9]+)\t+device.*', line): # We are an emulator
-                continue                                           # Skip it
-            m = re.search('^([A-z0-9.:]+.*?)\t+device.*', line)
+    try:
+        result = subprocess.check_output(f'"{PLATDIR}adb" devices', shell=True)
+        lines = result.splitlines()
+        for line in lines[1:]:
+            if line:
+                l = str(line, 'utf-8')  # convert byte to string
+                if emulator:
+                    m = re.search(r'^(emulator-\d+)\s+device.*', l)
+                else:
+                    if re.search(r'^(emulator-\d+)\s+device.*', l):  # We are an emulator
+                        continue  # Skip it
+                    m = re.search(r'^([\w\d]+)\s+device.*', l)
+                if m:
+                    break
         if m:
-            break
-    if m:
-        return m.group(1)
-    return False
+            return m.group(1)
+        return False
+    except subprocess.CalledProcessError as e:
+        print(f"Problem checking for devices : status {e.returncode}\n")
+        return False
+
 
 def killadb():
     """Time to nuke adb!"""
-    subprocess.check_output(PLATDIR + "commands-for-Appinventor/adb kill-server", shell=True, close_fds=True)
-    sys.stdout.write("Killed adb\n")
-    subprocess.check_output(PLATDIR + "commands-for-Appinventor/kill-emulator", shell=True, close_fds=True)
-    sys.stdout.write("Killed emulator\n")
+    try:
+        subprocess.check_output(f'"{PLATDIR}adb" kill-server', shell=True)
+        print("Killed adb\n")
+    except subprocess.CalledProcessError as e:
+        print(f"Problem stopping adb : status {e.returncode}\n")
+        return ''
+
+
+def killemulator():
+    try:
+        subprocess.check_output(f'"{PLATDIR}kill-emulator"', shell=True)
+        print("Killed emulator\n")
+    except subprocess.CalledProcessError as e:
+        print(f"Problem stopping emulator : status {e.returncode}\n")
+        return ''
+
 
 def shutdown():
-    try:                                # Be quiet...
-        killadb()
-    except:
-        pass
+    killadb()
+    killemulator()
```

Comment:
In version 3.0, the kill-emulator script calls adb emu kill, which I expect will just restart the server previously killed by killadb(). We should think about inverting the calls to killadb and killemulator so that the latter runs first.

Reply:
now it calls killemulator() before killadb()

Reply:
BTW, where is the source for version 3.0?